### ISSUE TITLE
Allow import of OWASP AntiSamy XML policy files

### ISSUE THREAD

**Issue by buckett** [OWASP/java-html-sanitizer#38]
To ease migration it would be useful if java-html-sanitizer could import AntiSamy XML policy files.

**Comment by jpetren** [comment]
I agree.  I'm looking for a way to preserve extra whitespaces and it looks like the only way to do that is to import a policy.

**Comment by jmanico** [comment]
Perhaps not "import" AntiSamy, but what about a standalone "convert" AntiSamy tool? That might be easier to support and would not require any enhancements in the current engine. Sound reasonable?

**Comment by efge** [comment]
@mikesamuel I plan on adding some code to do this. I would:

- take the AntiSamy policy model parsing from AntiSamy itself (in `org.owasp.validator.html.model.*` and `org.owasp.validator.html.Policy`) to avoid reinventing the parsing code,
- try to evaluate what we can from this policy and map it to the `HtmlPolicyBuilder` model, using a new `loadAntiSamyPolicy` method.

Does that plan work with you?

**Comment by jmanico** [comment]
This plan works for me. I’ve gotten this request from the field often. Thank you for taking this on Mr. Samuel.

Aloha,
--
Jim Manico
@Manicode
Secure Coding Education
+1 (808) 652-3805

**Comment by efge** [comment]
Please review the PR in #161

**Comment by efge** [comment]
Ping any committer?

**Comment by jmanico** [comment]
Thanks for the suggestion, and apologies that it took this long to get a reply.

I am going to close this as out of scope. Supporting AntiSamy's XML policy format would mean carrying a parser and a semantic mapping for another project's policy model inside the security-critical core of this one, and the two models do not line up cleanly. Where they disagree, the mismatch would surface as a silently weaker policy than the author intended — the worst failure mode for a sanitizer, and one that is very hard to notice in review.

Two alternatives that seem better to me:

- AntiSamy is still maintained as a separate OWASP project, so staying on it is entirely reasonable if its policy format is what you need.
- If you are migrating, a translation step that reads your AntiSamy XML and emits `HtmlPolicyBuilder` calls can live in your own build or application, where it is reviewable alongside the rest of your security configuration. Nothing in this library stands in the way of that.

Thanks again for the idea.
